### ISSUE TITLE
Interface Inferred from Property Not Rejecting Invalid Properties

### ISSUE THREAD

**Issue by MichaelPastuch** [microsoft/TypeScript#50732]
# Bug Report

### 🔎 Search Terms

"type infer", "conditional interface"

### 🕗 Version & Regression Information

- This is the behavior in every version I tried (4.2, 4.6, 4.8, and nightly)

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBAKgcgVwLY3BAzlAvFAjFAHygCZCoBmAbgChqBLAO2AgCcAzAQwGNoBJAIQ7poAb2oBIUJABcsRCjToaAX1qNm7bnwDKwFowDmUCAA9mDACaYBQ0RKkRZxGuPR7Dst-oYGVapqycPFC8ACIczMZmEJbWgsJQYpJoslQSFhGOUOHMftRcAPYMblDMbrI2CUS8ut5G1TnQOEkOqQA0El4eUABEZcA91Mo0QA)

### 💻 Code

```ts
type TNumTypes = 1 | 2 | 3;

interface IBase {
	type: TNumTypes;
}

interface IString extends IBase {
	type: 2;
	string: string;
}

interface IDate extends IBase {
	type: 3;
	date: Date;
}

const test: IBase | IString | IDate = {
	type: 3,
	string: "test"
};
```

### 🙁 Actual behavior

The `test` object with a `type` value of `2` should flag the `string` property as invalid, since it is not in `IBase` or `IDate`, but the object is considered valid.

### 🙂 Expected behavior

The `test` object should infer its interface from the value of `type`. Any valid value could be an `IBase`, but a value of `2` should not be an `IDate` and a value of `3` should not be an `IString`.

**Comment by MartinJohns** [comment]
Duplicate of #20863. Your object is a valid `IBase`. Additional properties are allowed.

**Comment by fatcerberus** [comment]
IIRC this is a limitation of excess property checking; if the property exists in *any* of the union constituents, it isn't flagged.  Note that excess property errors are not type errors *per se*; TypeScript doesn't support [exact types](https://github.com/microsoft/TypeScript/issues/12936).

**Comment by MichaelPastuch** [comment]
Thank you for the swift responses. I am happy with the explanation and can work around my issue.
